The build failed on Linux with the following error:

``` bash
$ swift test
Compiling Swift Module 'Himotokitest' (6 sources)
/home/travis/build/ikesyo/Himotoki/Tests/Himotoki/DecodeErrorTest.swift:25:21: error: constructing an object of class type 'Self' with a metatype value must use a 'required' initializer
        return self.init(string: value)!
               ~~~~ ^
Foundation.NSURL:15:24: note: selected non-required initializer 'init(string:)'
    public convenience init?(string URLString: String)
                       ^
<unknown>:0: error: build had 1 command failures
error: exit(1): swift-build-tool -f /home/travis/build/ikesyo/Himotoki/.build/debug.yaml test
```

@tarunon Do you have any solution or workaround for this?

--- a/Sources/Decodable.swift
+++ b/Sources/Decodable.swift
@@ -7,8 +7,6 @@
 //
 
 public protocol Decodable {
-    typealias DecodedType = Self
-
     /// - Throws: DecodeError
-    static func decode(e: Extractor) throws -> DecodedType
+    static func decode(e: Extractor) throws -> Self
 }

--- a/Sources/Extractor.swift
+++ b/Sources/Extractor.swift
@@ -39,7 +39,7 @@ public struct Extractor {
     }
 
     /// - Throws: DecodeError
-    public func value<T: Decodable where T.DecodedType == T>(keyPath: KeyPath) throws -> T {
+    public func value<T: Decodable>(keyPath: KeyPath) throws -> T {
         guard let rawValue = try _rawValue(keyPath) else {
             throw DecodeError.MissingKeyPath(keyPath)
         }
@@ -54,7 +54,7 @@ public struct Extractor {
     }
 
     /// - Throws: DecodeError
-    public func valueOptional<T: Decodable where T.DecodedType == T>(keyPath: KeyPath) throws -> T? {
+    public func valueOptional<T: Decodable>(keyPath: KeyPath) throws -> T? {
         do {
             return try value(keyPath) as T
         } catch let DecodeError.MissingKeyPath(missing) where missing == keyPath {
@@ -63,7 +63,7 @@ public struct Extractor {
     }
 
     /// - Throws: DecodeError
-    public func array<T: Decodable where T.DecodedType == T>(keyPath: KeyPath) throws -> [T] {
+    public func array<T: Decodable>(keyPath: KeyPath) throws -> [T] {
         guard let array: [T] = try arrayOptional(keyPath) else {
             throw DecodeError.MissingKeyPath(keyPath)
         }
@@ -72,12 +72,12 @@ public struct Extractor {
     }
 
     /// - Throws: DecodeError
-    public func arrayOptional<T: Decodable where T.DecodedType == T>(keyPath: KeyPath) throws -> [T]? {
+    public func arrayOptional<T: Decodable>(keyPath: KeyPath) throws -> [T]? {
         return try _rawValue(keyPath).map(decodeArray)
     }
 
     /// - Throws: DecodeError
-    public func dictionary<T: Decodable where T.DecodedType == T>(keyPath: KeyPath) throws -> [String: T] {
+    public func dictionary<T: Decodable>(keyPath: KeyPath) throws -> [String: T] {
         guard let dictionary: [String: T] = try dictionaryOptional(keyPath) else {
             throw DecodeError.MissingKeyPath(keyPath)
         }
@@ -86,7 +86,7 @@ public struct Extractor {
     }
 
     /// - Throws: DecodeError
-    public func dictionaryOptional<T: Decodable where T.DecodedType == T>(keyPath: KeyPath) throws -> [String: T]? {
+    public func dictionaryOptional<T: Decodable>(keyPath: KeyPath) throws -> [String: T]? {
         return try _rawValue(keyPath).map(decodeDictionary)
     }
 }

--- a/Sources/NSNumber.swift
+++ b/Sources/NSNumber.swift
@@ -9,7 +9,7 @@
 import class Foundation.NSNumber
 
 extension NSNumber: Decodable {
-    public static func decode(e: Extractor) throws -> NSNumber {
+    public static func decode(e: Extractor) throws -> Self {
         return try castOrFail(e)
     }
 }

--- a/Sources/Operators.swift
+++ b/Sources/Operators.swift
@@ -14,31 +14,31 @@ infix operator <|-| { associativity left precedence 150 }
 infix operator <|-|? { associativity left precedence 150 }
 
 /// - Throws: DecodeError
-public func <| <T: Decodable where T.DecodedType == T>(e: Extractor, keyPath: KeyPath) throws -> T {
+public func <| <T: Decodable>(e: Extractor, keyPath: KeyPath) throws -> T {
     return try e.value(keyPath)
 }
 
 /// - Throws: DecodeError
-public func <|? <T: Decodable where T.DecodedType == T>(e: Extractor, keyPath: KeyPath) throws -> T? {
+public func <|? <T: Decodable>(e: Extractor, keyPath: KeyPath) throws -> T? {
     return try e.valueOptional(keyPath)
 }
 
 /// - Throws: DecodeError
-public func <|| <T: Decodable where T.DecodedType == T>(e: Extractor, keyPath: KeyPath) throws -> [T] {
+public func <|| <T: Decodable>(e: Extractor, keyPath: KeyPath) throws -> [T] {
     return try e.array(keyPath)
 }
 
 /// - Throws: DecodeError
-public func <||? <T: Decodable where T.DecodedType == T>(e: Extractor, keyPath: KeyPath) throws -> [T]? {
+public func <||? <T: Decodable>(e: Extractor, keyPath: KeyPath) throws -> [T]? {
     return try e.arrayOptional(keyPath)
 }
 
 /// - Throws: DecodeError
-public func <|-| <T: Decodable where T.DecodedType == T>(e: Extractor, keyPath: KeyPath) throws -> [String: T] {
+public func <|-| <T: Decodable>(e: Extractor, keyPath: KeyPath) throws -> [String: T] {
     return try e.dictionary(keyPath)
 }
 
 /// - Throws: DecodeError
-public func <|-|? <T: Decodable where T.DecodedType == T>(e: Extractor, keyPath: KeyPath) throws -> [String: T]? {
+public func <|-|? <T: Decodable>(e: Extractor, keyPath: KeyPath) throws -> [String: T]? {
     return try e.dictionaryOptional(keyPath)
 }

--- a/Sources/RawRepresentable.swift
+++ b/Sources/RawRepresentable.swift
@@ -6,7 +6,7 @@
 //  Copyright © 2015 Syo Ikeda. All rights reserved.
 //
 
-public extension RawRepresentable where Self: Decodable, Self.DecodedType == Self, RawValue: Decodable, RawValue.DecodedType == RawValue {
+public extension RawRepresentable where Self: Decodable, RawValue: Decodable {
     static func decode(e: Extractor) throws -> Self {
         let rawValue = try RawValue.decode(e)
 

--- a/Sources/decode.swift
+++ b/Sources/decode.swift
@@ -7,18 +7,18 @@
 //
 
 /// - Throws: DecodeError
-public func decodeValue<T: Decodable where T.DecodedType == T>(JSON: AnyJSON) throws -> T {
+public func decodeValue<T: Decodable>(JSON: AnyJSON) throws -> T {
     let extractor = Extractor(JSON)
     return try T.decode(extractor)
 }
 
 /// - Throws: DecodeError
-public func decodeValue<T: Decodable where T.DecodedType == T>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> T {
+public func decodeValue<T: Decodable>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> T {
     return try decodeValue(JSON) <| rootKeyPath
 }
 
 /// - Throws: DecodeError
-public func decodeArray<T: Decodable where T.DecodedType == T>(JSON: AnyJSON) throws -> [T] {
+public func decodeArray<T: Decodable>(JSON: AnyJSON) throws -> [T] {
     guard let array = JSON as? [AnyJSON] else {
         throw typeMismatch("Array", actual: JSON, keyPath: nil)
     }
@@ -27,12 +27,12 @@ public func decodeArray<T: Decodable where T.DecodedType == T>(JSON: AnyJSON) th
 }
 
 /// - Throws: DecodeError
-public func decodeArray<T: Decodable where T.DecodedType == T>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> [T] {
+public func decodeArray<T: Decodable>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> [T] {
     return try decodeValue(JSON) <|| rootKeyPath
 }
 
 /// - Throws: DecodeError
-public func decodeDictionary<T: Decodable where T.DecodedType == T>(JSON: AnyJSON) throws -> [String: T] {
+public func decodeDictionary<T: Decodable>(JSON: AnyJSON) throws -> [String: T] {
     guard let dictionary = JSON as? [String: AnyJSON] else {
         throw typeMismatch("Dictionary", actual: JSON, keyPath: nil)
     }
@@ -45,7 +45,7 @@ public func decodeDictionary<T: Decodable where T.DecodedType == T>(JSON: AnyJSO
 }
 
 /// - Throws: DecodeError
-public func decodeDictionary<T: Decodable where T.DecodedType == T>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> [String: T] {
+public func decodeDictionary<T: Decodable>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> [String: T] {
     return try decodeValue(JSON) <|-| rootKeyPath
 }
 
@@ -53,12 +53,12 @@ public func decodeDictionary<T: Decodable where T.DecodedType == T>(JSON: AnyJSO
 
 /// - Throws: DecodeError
 @available(*, deprecated, renamed="decodeValue")
-public func decode<T: Decodable where T.DecodedType == T>(JSON: AnyJSON) throws -> T {
+public func decode<T: Decodable>(JSON: AnyJSON) throws -> T {
     return try decodeValue(JSON)
 }
 
 /// - Throws: DecodeError
 @available(*, deprecated, renamed="decodeValue")
-public func decode<T: Decodable where T.DecodedType == T>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> T {
+public func decode<T: Decodable>(JSON: AnyJSON, rootKeyPath: KeyPath) throws -> T {
     return try decodeValue(JSON)
 }

--- a/Tests/Himotoki/DecodeErrorTest.swift
+++ b/Tests/Himotoki/DecodeErrorTest.swift
@@ -28,7 +28,7 @@ extension NSURL: Decodable {
         if value.hasPrefix("file://") {
             throw customError("File URL is not supported")
         }
-        return try castOrFail(self.init(string: value))
+        return try castOrFail(NSURL(string: value))
     }
 }
 

--- a/Tests/Himotoki/DecodeErrorTest.swift
+++ b/Tests/Himotoki/DecodeErrorTest.swift
@@ -10,6 +10,13 @@ import Foundation
 import XCTest
 import Himotoki
 
+func castOrFail<T>(any: Any?) throws -> T {
+    guard let casted = any as? T else {
+        throw typeMismatch("\(T.self)", actual: any, keyPath: nil)
+    }
+    return casted
+}
+
 extension NSURL: Decodable {
     public static func decode(e: Extractor) throws -> Self {
         let value = try String.decode(e)
@@ -21,8 +28,7 @@ extension NSURL: Decodable {
         if value.hasPrefix("file://") {
             throw customError("File URL is not supported")
         }
-
-        return self.init(string: value)!
+        return try castOrFail(self.init(string: value))
     }
 }
 

--- a/Tests/Himotoki/DecodeErrorTest.swift
+++ b/Tests/Himotoki/DecodeErrorTest.swift
@@ -11,7 +11,7 @@ import XCTest
 import Himotoki
 
 extension NSURL: Decodable {
-    public static func decode(e: Extractor) throws -> NSURL {
+    public static func decode(e: Extractor) throws -> Self {
         let value = try String.decode(e)
 
         if value.isEmpty {
@@ -22,7 +22,7 @@ extension NSURL: Decodable {
             throw customError("File URL is not supported")
         }
 
-        return NSURL(string: value)!
+        return self.init(string: value)!
     }
 }
 

